mifospay/src/main/java/org/mifospay/receipt/presenter/ReceiptViewModel.kt

```diff
@@ -0,0 +1,124 @@
+            object : UseCase.UseCaseCallback<DownloadTransactionReceipt.ResponseValue> {
+                override fun onSuccess(response: DownloadTransactionReceipt.ResponseValue) {
+                    val filename = Constants.RECEIPT + transactionId + Constants.PDF
+                    writeReceiptToPDF(response.responseBody, filename)
+                }
+
+                override fun onError(message: String) {
+                    _receiptState.value = ReceiptUiState.Error(message)
+                }
+            })
+    }
+
+    fun writeReceiptToPDF(responseBody: ResponseBody?, filename: String) {
+        val mifosDirectory = File(
+            Environment.getExternalStorageDirectory(),
+            Constants.MIFOSPAY
+        )
+        if (!mifosDirectory.exists()) {
+            mifosDirectory.mkdirs()
+        }
+        val documentFile = File(mifosDirectory.path, filename)
+        if (FileUtils.writeInputStreamDataToFile(responseBody!!.byteStream(), documentFile)) {
+            _fileState.value = PassFileState(documentFile, true)
+        } else {
+            _fileState.update { currentState ->
+                currentState.copy(writeReceiptToPDFisSuccess = false)
+            }
+        }
+    }
+
+    fun fetchTransaction(transactionId: String?) {
+        val accountId = preferencesHelper.accountId
+
+        if (transactionId != null) {
+            mUseCaseHandler.execute(fetchAccountTransactionUseCase,
+                FetchAccountTransaction.RequestValues(accountId, transactionId.toLong()),
+                object : UseCase.UseCaseCallback<FetchAccountTransaction.ResponseValue> {
+                    override fun onSuccess(response: FetchAccountTransaction.ResponseValue) {
+                        _receiptState.value = ReceiptUiState.Success(response.transaction)
+                        fetchTransfer(response.transaction.transferId)
+                    }
+
+                    override fun onError(message: String) {
+                        if (message == Constants.UNAUTHORIZED_ERROR) {
+                            _receiptState.value = ReceiptUiState.OpenPassCodeActivity
+                        } else {
+                            _receiptState.value = ReceiptUiState.Error(message)
+                        }
+                    }
+                }
+            )
+        }
+    }
+
+    fun fetchTransfer(transferId: Long) {
+        mUseCaseHandler.execute(fetchAccountTransferUseCase,
+            FetchAccountTransfer.RequestValues(transferId),
+            object : UseCase.UseCaseCallback<FetchAccountTransfer.ResponseValue?> {
+                override fun onSuccess(response: FetchAccountTransfer.ResponseValue?) {}
+
+                override fun onError(message: String) {
+                    _receiptState.value = ReceiptUiState.Error(message)
+                }
+            })
+    }
+}
+
+data class PassFileState(
+    val file: File = File(" "),
+    val writeReceiptToPDFisSuccess: Boolean = false
+)
```

Comment:
interface -> data class, success object, error object?

Reply:
if documentFile isn't available then 'writeReceiptToPDFisSuccess' will become false and will handle the error as follows:
![image](https://github.com/openMF/mobile-wallet/assets/95978829/35f9d274-335f-43a3-ae14-e3f29e1c6290)